sw/device/tests/entropy_src_fw_observe_many_contiguous.c

```diff
@@ -0,0 +1,240 @@
+   */
+  kBitsPerSample = 4,
+  /**
+   * Size of buffer in words to hold all the samples, assuming
+   * 4-bit samples at the most.
+   */
+  kFifoBufferSizeWords =
+      kContiguousSamplesCount * kBitsPerSample / sizeof(uint32_t),
+};
+
+static uint32_t sample_buffer[kFifoBufferSizeWords];
+static dif_entropy_src_t entropy_src;
+static dif_csrng_t csrng;
+static dif_edn_t edn0;
+static dif_edn_t edn1;
+
+/**
+ * Determine whether the observe FIFO has overflowed.
+ *
+ * FIXME Normally, one would rely on the FW_OV_RD_FIFO_OVERFLOW
+ * register but due to an RTL bug, the overflow bit is pulsed
+ * instead of latched so we cannot rely on it. Instead, rely
+ * on OBSERVE_FIFO_DEPTH and assume that if the FIFO is full
+ * then it has overflowed.
+ */
+bool entropy_src_fifo_has_overflowed(void) {
+  uint32_t fifo_depth;
+  CHECK_DIF_OK(dif_entropy_src_get_fifo_depth(&entropy_src, &fifo_depth));
+  return fifo_depth == ENTROPY_SRC_PARAM_OBSERVE_FIFO_DEPTH;
+}
+
+/**
+ * Drain observe FIFO and clear overflow status if set.
+ */
+static void drain_observe_fifo(void) {
+  // This value is arbitrary, it could be 1 but since there is some
+  // overhead in dif_entropy_src_observe_fifo_nonblocking_read, it's better
+  // to read several words every time to drain the FIFO quickly.
+  const size_t kDrainCount = 32;
+  size_t len;
+  // Read from the FIFO until we get a short read which means that the FIFO was
+  // emptied.
+  LOG_INFO("drain observe FIFO overflow...");
+  do {
+    len = kDrainCount;
+    CHECK_DIF_OK(dif_entropy_src_observe_fifo_nonblocking_read(&entropy_src,
+                                                               NULL, &len));
+  } while (len == kDrainCount);
+  CHECK_DIF_OK(dif_entropy_src_clear_fifo_overflow(&entropy_src));
+}
+
+/**
+ * Let observe FIFO overflow.
+ */
+static void let_observe_fifo_overflow(uint32_t timeout_usec) {
+  LOG_INFO("let observe FIFO overflow...");
+  ibex_timeout_t tmo = ibex_timeout_init(timeout_usec);
+  while (!entropy_src_fifo_has_overflowed() && !ibex_timeout_check(&tmo)) {
+  }
+  CHECK(!ibex_timeout_check(&tmo), "observe FIFO did not overflow in time");
+  uint64_t elapsed = ibex_timeout_elapsed(&tmo);
+  LOG_INFO("done in %ums", (uint32_t)udiv64_slow(elapsed, 1000, NULL));
+}
+
+// Configure the entropy complex.
+static void entropy_config(dif_entropy_src_single_bit_mode_t single_bit_mode) {
+  dif_edn_auto_params_t edn_params0 =
+      edn_testutils_auto_params_build(false, /*res_itval=*/0, /*glen_val=*/0);
+  dif_edn_auto_params_t edn_params1 =
+      edn_testutils_auto_params_build(false, /*res_itval=*/0, /*glen_val=*/0);
+  // Disable the entropy complex.
+  CHECK_STATUS_OK(entropy_testutils_stop_all());
+
+  // Disable all health tests.
+  CHECK_DIF_OK(dif_entropy_src_health_test_configure(
+      &entropy_src, (dif_entropy_src_health_test_config_t){
+                        .test_type = kDifEntropySrcTestRepetitionCount,
+                        .high_threshold = 0xffffffff,
```

Comment:
Nit
```suggestion
                        .high_threshold = UINT32_MAX,
```

Reply:
This is not really a single uint32_t, it's the raw encoding of a register that has two 16-bit fields. Unfortunately all of the entropy_src code has constants like these which is not ideal.